pkg/cdi/cache.go

```diff
@@ -0,0 +1,551 @@
+
+// CacheOption is an option for a Cache.
+type CacheOption func(*Cache) error
+
+// WithSpecDirs overrides the default directories CDI Specs are read from.
+func WithSpecDirs(dirs ...string) CacheOption {
+	return func(c *Cache) error {
+		c.specDirs = make([]string, len(dirs))
+		copy(c.specDirs, dirs)
+		return nil
+	}
+}
+
+// WithRefresh specifies whether updated Spec files are reloaded.
+func WithRefresh(refresh bool) CacheOption {
+	return func(c *Cache) error {
+		c.refresh = refresh
+		return nil
+	}
+}
+
+// WithFilter specifies a filter for loading devices from Specs.
+func WithFilter(f Filter) CacheOption {
+	return func(c *Cache) error {
+		c.filter = &f
+		return nil
+	}
+}
+
+// Cache CDI Specs loaded from Spec directories.
+type Cache struct {
+	sync.Mutex
+	filter   *Filter
+	refresh  bool
+	blocked  uint
+	pending  bool
+	w        *fsnotify.Watcher
+	specDirs []string
+	specs    map[string][]*Spec
+	errors   []error
+	devices  map[string]*Device
+}
+
+// NewCache creates a new CDI Cache. The cache is populated from the CDI Spec
+// directories which can be specified using the WithSpecDirs options. If no
+// directories are gives, the defaults DefaultStaticDir, DefaultDynamicDir
+// ones are used.
+func NewCache(opts ...CacheOption) (*Cache, error) {
+	c := &Cache{
+		specDirs: []string{DefaultStaticDir, DefaultDynamicDir},
+	}
+
+	for _, o := range opts {
+		if err := o(c); err != nil {
+			return nil, errors.Wrap(err, "failed create cache, failed option")
+		}
+	}
+
+	//c.watch()
+
+	err := c.Refresh()
+
+	return c, err
+}
+
+// GetSpecDirectories return the spec directories for this cache.
+func (c *Cache) GetSpecDirectories() []string {
+	if c == nil {
+		return nil
+	}
+	dirs := make([]string, len(c.specDirs))
+	copy(dirs, c.specDirs)
+	return dirs
+}
+
+// ListVendors lists all vendors known to the cache.
+func (c *Cache) ListVendors() []string {
+	if c == nil {
+		return nil
+	}
+
+	c.Lock()
+	defer c.Unlock()
+
+	var vendors []string
+
+	for vendor := range c.specs {
+		vendors = append(vendors, vendor)
+	}
+
+	sort.Strings(vendors)
+
+	return vendors
+}
+
+// GetVendorSpecs returns all CDI Specs for a vendor.
+func (c *Cache) GetVendorSpecs(vendor string) []*Spec {
+	if c == nil {
+		return nil
+	}
+
+	c.Lock()
+	defer c.Unlock()
+
+	return c.specs[vendor]
+}
+
+// HasDevice checks if the cache contains the qualified device.
+func (c *Cache) HasDevice(device string) bool {
+	if c == nil {
+		return false
+	}
+
+	c.Lock()
+	defer c.Unlock()
+
+	_, found := c.devices[device]
+	return found
+}
+
+// ListDevices lists all qualified devices matching a filter.
+func (c *Cache) ListDevices(f *Filter) ([]string, error) {
+	if c == nil {
+		return nil, nil
+	}
+
+	c.Lock()
+	defer c.Unlock()
+
+	var devices []string
+	for device, d := range c.devices {
+		if f.Match(d.GetSpec().GetVendor(), d.GetSpec().GetClass(), d.Name) {
+			devices = append(devices, device)
+		}
+	}
+	sort.Strings(devices)
```

Comment:
No. But I think it is a good idea if the package to provides a stable answer to a question. IOW, for the same set of CDI Specs always produce the same devices in the same order. Since there are maps involved, sorting is necessary to achieve that.

Reply:
Yes, stable responses are good. I was wondering about the time required to process large lists of devices or do we not consider this to be a problem?

Reply:
Yes, we definitely care about performance. That's part of the reason we try to update the implementation. If there is a concern about the performance impact of sorting here, then I think we should quantify and measure. So we need to state the order of magnitude for a device set CDI is expected to reasonably handle (hundreds, thousands, tens of thousands?) and then measure the cost of sorting such a slice of reasonable long qualified device-ish strings. Then we can make an educated guess if the concern is valid or not.

Reply:
Let's leave it sorted for now. If we see performance becoming a problem, we can revisit this. It may not be required to optimise this at all.

Reply:
Yes. I did some very simple measurements of potential perf-impact of both sorting and globbing. Sorting a randomly generated set (of a particular size) of fully qualified device names of a particular length look pretty cheap and hence safe to do always. For instance, sorting a set of 128 device IDs of 64 bytes (on my laptop) takes 8.17 usec's. I think it's more realistic to have something like 64 devices of max. 32 length, which would take 3.86 usec's. So I left that in...

The golang stock globbing implementation is more expensive and scales worse with set size. The real problem with it IMO is that it does not support alternatives (IOW, you can't do something like `echo test-oci.spec | cdi inject - "vendor.com/{gpu,fpga}/*"` and expect it to work. So I took that one out completely. It is anyway trivial to do glob-based filtering on the CDI caller/client side, so this should not be a bad biggie ATM.